module/zfs/arc.c

```diff
@@ -5563,6 +5563,12 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 		if (HDR_IO_IN_PROGRESS(hdr)) {
 			zio_t *head_zio = hdr->b_l1hdr.b_acb->acb_zio_head;
 
+			if (*arc_flags & ARC_FLAG_CACHED_ONLY) {
+				mutex_exit(hash_lock);
```

Comment:
Is `hash_lock` guaranteed to be not `NULL` here? All other places seem to check this?

Reply:
Yes, it is guaranteed to not be NULL.  It comes from `buf_hash_find()`, which always fills it in if a hdr is found (`hdr != NULL`).  I'm not sure which other places you are thinking of, but the same thing is done throughout this code block, see lines 5591, 5616, 5619, 5689.  In the "else" clause on line 5697, the hdr might or might not have been found, and thus the hash_lock might or might not be filled in, so in that block we need to check.

Reply:
@ahrens Ah yeah, that makes sense. Thanks for the explanation :bow:.